ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Decision.hs

```diff
@@ -0,0 +1,345 @@
+    objectsExpectedSoonMultiplicities :: Map objectId ObjectMultiplicity
+    objectsExpectedSoonMultiplicities = Map.unionWith (+) objectsInFlightMultiplicities objectsOwtPoolMultiplicities
+
+    -- Now we join objectsToSortedProviders and objectsExpectedSoonMultiplicities maps on objectId for easy fold
+    objectsToProvidersAndExpectedMultiplicities ::
+      Map objectId ([(peerAddr, NumObjectsReq)], ObjectMultiplicity)
+    objectsToProvidersAndExpectedMultiplicities =
+      Map.merge
+        -- if an objectId is missing from objectsExpectedSoonMultiplicities, then its expected multiplicity is 0
+        (Map.mapMissing \_ providers -> (providers, 0))
+        -- if an objectId is missing from objectsToSortedProviders, then we don't care about it
+        Map.dropMissing
+        -- Combine in a tuple the list of providers and the expected multiplicity
+        (Map.zipWithMatched \_ providers expectedMultiplicity -> (providers, expectedMultiplicity))
+        objectsToSortedProviders
+        objectsExpectedSoonMultiplicities
+
+    -- NOW HERE TAKE PLACE THE ACTUAL DECISION LOGIC AND ATTRIBUTION OF OBJECTS TO PEERS
+
+    -- The current decision logic is greedy on objects, so it will try to request as many copies of the same object as possible,
+    -- meaning we will have optimal coverage of the first objects, but might not request some other objects at all if they are (only) provided by peers that are already saturated.
+
+    -- Now we compute the actual attribution of downloads for peers
+    DownloadPickState{peersToObjectsToReq} =
+      -- We iterate over each objectId and the corresponding (providers, expectedMultiplicity)
+      Map.foldlWithKey'
+        ( \st objectId (providers, expectedMultiplicity) ->
+            -- reset the objectMultiplicity counter for each new objectId
+            let st' = st{objectMultiplicity = 0}
+             in -- We iterate over the list of providers, and pick them or not according to the current state
+                -- When a peer is selected as a provider for this objectId, we insert the objectId in the peer's set in peersToObjectsToReq (inside St)
+                -- So the result of the filtering of providers is part of the final St state
+                Foldable.foldl'
+                  (howToFoldProviders objectId expectedMultiplicity)
+                  st'
+                  providers
+        )
+        DownloadPickState
+          { totalNumObjectsToReq = 0
+          , objectMultiplicity = 0
+          , peersToObjectsToReq = Map.empty
+          }
+        objectsToProvidersAndExpectedMultiplicities
+
+    -- This function decides whether or not we should select a given peer as provider for the current objectId
+    -- it takes into account if we are expecting to obtain the object from other sources (either inflight/owt pool already, or if the object will be requested from already selected peers in this given round)
+    howToFoldProviders ::
```

Comment:
I would rename this function to something depicting its purpose, e.g. `pickProviderForObjectId`